### ISSUE TITLE
Pod never leaves pending.

### ISSUE THREAD

**Issue by empath-75** [deislabs/krustlet#165]
I'm not sure what information you need to troubleshoot, but i'm running krustlet-wasi.  I've got the node in a ready state and the pod from the walk through is scheduled on the node, but it never leaves pending state.. the logs for krustlet aren't particularly great -- they don't show anything at all unless i delete the pod or edit it:

if i edit the pod, it shows this:

```
Apr 07 21:00:28 ip-10-205-123-180.ec2.internal krustlet-wasi[23351]: [2020-04-07T21:00:28Z INFO  wasi_provider] Pod modified
Apr 07 21:00:28 ip-10-205-123-180.ec2.internal krustlet-wasi[23351]: [2020-04-07T21:00:28Z INFO  wasi_provider] Modified pod spec: PodStatu
Apr 07 21:00:28 ip-10-205-123-180.ec2.internal krustlet-wasi[23351]:         conditions: Some(
Apr 07 21:00:28 ip-10-205-123-180.ec2.internal krustlet-wasi[23351]:             [
Apr 07 21:00:28 ip-10-205-123-180.ec2.internal krustlet-wasi[23351]:                 PodCondition {
Apr 07 21:00:28 ip-10-205-123-180.ec2.internal krustlet-wasi[23351]:                     last_probe_time: None,
Apr 07 21:00:28 ip-10-205-123-180.ec2.internal krustlet-wasi[23351]:                     last_transition_time: Some(
Apr 07 21:00:28 ip-10-205-123-180.ec2.internal krustlet-wasi[23351]:                         Time(
Apr 07 21:00:28 ip-10-205-123-180.ec2.internal krustlet-wasi[23351]:                             2020-04-07T20:58:04Z,
Apr 07 21:00:28 ip-10-205-123-180.ec2.internal krustlet-wasi[23351]:                         ),
Apr 07 21:00:28 ip-10-205-123-180.ec2.internal krustlet-wasi[23351]:                     ),
Apr 07 21:00:28 ip-10-205-123-180.ec2.internal krustlet-wasi[23351]:                     message: None,
Apr 07 21:00:28 ip-10-205-123-180.ec2.internal krustlet-wasi[23351]:                     reason: None,
Apr 07 21:00:28 ip-10-205-123-180.ec2.internal krustlet-wasi[23351]:                     status: "True",
Apr 07 21:00:28 ip-10-205-123-180.ec2.internal krustlet-wasi[23351]:                     type_: "PodScheduled",
Apr 07 21:00:28 ip-10-205-123-180.ec2.internal krustlet-wasi[23351]:                 },
Apr 07 21:00:28 ip-10-205-123-180.ec2.internal krustlet-wasi[23351]:             ],
Apr 07 21:00:28 ip-10-205-123-180.ec2.internal krustlet-wasi[23351]:         ),
Apr 07 21:00:28 ip-10-205-123-180.ec2.internal krustlet-wasi[23351]:         container_statuses: None,
Apr 07 21:00:28 ip-10-205-123-180.ec2.internal krustlet-wasi[23351]:         ephemeral_container_statuses: None,
Apr 07 21:00:28 ip-10-205-123-180.ec2.internal krustlet-wasi[23351]:         host_ip: None,
Apr 07 21:00:28 ip-10-205-123-180.ec2.internal krustlet-wasi[23351]:         init_container_statuses: None,
Apr 07 21:00:28 ip-10-205-123-180.ec2.internal krustlet-wasi[23351]:         message: None,
Apr 07 21:00:28 ip-10-205-123-180.ec2.internal krustlet-wasi[23351]:         nominated_node_name: None,
Apr 07 21:00:28 ip-10-205-123-180.ec2.internal krustlet-wasi[23351]:         phase: Some(
Apr 07 21:00:28 ip-10-205-123-180.ec2.internal krustlet-wasi[23351]:             "Pending",
Apr 07 21:00:28 ip-10-205-123-180.ec2.internal krustlet-wasi[23351]:         ),
Apr 07 21:00:28 ip-10-205-123-180.ec2.internal krustlet-wasi[23351]:         pod_ip: None,
Apr 07 21:00:28 ip-10-205-123-180.ec2.internal krustlet-wasi[23351]:         pod_ips: None,
Apr 07 21:00:28 ip-10-205-123-180.ec2.internal krustlet-wasi[23351]:         qos_class: Some(
Apr 07 21:00:28 ip-10-205-123-180.ec2.internal krustlet-wasi[23351]:             "BestEffort",
Apr 07 21:00:28 ip-10-205-123-180.ec2.internal krustlet-wasi[23351]:         ),
Apr 07 21:00:28 ip-10-205-123-180.ec2.internal krustlet-wasi[23351]:         reason: None,
Apr 07 21:00:28 ip-10-205-123-180.ec2.internal krustlet-wasi[23351]:         start_time: None,
Apr 07 21:00:28 ip-10-205-123-180.ec2.internal krustlet-wasi[23351]:     }
```

other than that, 

```
Apr 07 20:57:41 ip-10-205-123-180.ec2.internal systemd[1]: Started Krustlet, a kubelet implementation for running WASM.
```

is the only thing in the log.

**Comment by bacongobbler** [comment]
Hi @empath-75! Thanks for filing a ticket. Can you show what instructions you followed to reproduce your error?

If you can try to reproduce the issue using minikube or KinD, that would be great. That way we can try to replicate the issue and nail down the cause.

**Comment by empath-75** [comment]
I'm running on a kubeadm cluster in aws, so I sort of did my best at adapting all the other instructions and also bound the service account to system:node based on another issue I found.  I'll try and reproduce with KinD, and update this issue.

Is there a walk through of what is *supposed* to be happening after the pod is scheduled, or any way I can get more detailed logs?  (Would I see anything in the logs if it couldn't reach the repository, for example?)

**Comment by empath-75** [comment]
I have the same problem using kind.

I'm using the docker registry that I set up myself and I suspect the problem might be pulling from the registry, although i can push and pull from `wasm-to-oci`, but I can't figure out what is happening because there are no logs in krustlet.

```apiVersion: v1
kind: Pod
metadata:
  name: krustlet-tutorial
spec:
  containers:
    - name: krustlet-tutorial
      image: <docker registry>/demo:v1
  tolerations:
    - key: "krustlet/arch"
      operator: "Equal"
      value: "wasm32-wasi"
      effect: "NoExecute"
  nodeSelector:
    kubernetes.io/arch: wasm32-wasi
```

```
wasm-to-oci pull <docker registry>/demo:v1
INFO[0006] Pulled: <docker registry>/demo:v1
INFO[0006] Size: 2004905
INFO[0006] Digest: sha256:78373f5449a2dcead7359b826f51916485b347aa7e2c744695db488a3dd3c9b4
```

**Comment by empath-75** [comment]
i do have this in the krustlet logs which is related to proxy problems, probably -- it seems to ignore NO_PROXY/no_proxy, so i had to disable my proxy settings entirely for it to connect to the kind cluster.:

```[2020-04-08T15:46:49Z ERROR kubelet::kubelet] Error handling event: error sending request for url (https://k8s.gcr.io/v2/): error trying to connect: connection closed via error
```

**Comment by bacongobbler** [comment]
interesting findings. Thanks for sharing.

I don't think we have experience working with proxied environments. Do you have any PRs or suggestsions where we could make this workflow better?

**Comment by bacongobbler** [comment]
Were you able to resolve the original issue you were describing? After re-reading through this thread it's unclear whether you resolved the original issue described.

**Comment by empath-75** [comment]
No.  I still can't get it to work.  

What i'd really like is more detailed logging.  Just some recognition that it can see a pod, that it is attempting to download an image, etc.  I could do more useful troubleshooting or bug reporting if i knew what exactly wasn't working.

**Comment by bacongobbler** [comment]
Fair enough. We’d really appreciate a PR! If not we can get back to this at some point.

**Comment by empath-75** [comment]
After a lot of ugly println!() debugging, i fixed a few issues.. i've got some restrictions on submitting PRs from work that means it'll take a while before i submit a pr, but i can describe what i did, basically.. 

1) artifactory returns a `token` field, not `access_token`, so i added an alias.
2) details is optional in OciErrors (at least for docker registry)
3) the accept header needs to include `application/vnd.oci.image.manifest.v1+json` for docker registery to send the oci image.

Further, it would be nice to have some debug level logging that returns the text of reqwest responses before trying to parse it to json, because your structs might be throwing errors when you try to deserialize it.

**Comment by bacongobbler** [comment]
Thank you for looking into this!

From the sounds of it, it sounds like a lot of the issues you're finding here stems from the fact that a particular OCI registry (artifactory) is hard-coded to only support OCI images (i.e. Docker images), and it cannot accept alternative file formats (WebAssembly modules) as described in the [distribution specification](https://github.com/opencontainers/distribution-spec/).

So it sounds like there are a few action items to follow up here:

- we need to come up with some compatibility matrix for what containers registries work with [wasm-to-oci](https://github.com/engineerd/wasm-to-oci) and what don't
- we need to provide a way to gain more insight into reqwest, possibly by providing a `-vvv` flag to krustlet

Does that sound correct to you?

**Comment by rylev** [comment]
@empath-75 #186 should help a tiny bit with logging. FYI: setting the RUST_LOG environment variable gives more verbose logs. If you set it to `debug` you should see more information

**Comment by empath-75** [comment]
To clarify, the docker registry I'm using is this:

https://hub.docker.com/_/registry

Which from what i can tell, uses `distribution`.  That was what required the change to the 'accept' header in the image pull request, and which did not include a 'details' field in the error when that header was not included.

artifactory doesn't support oci images, and was only in the picture because krustlet was attempting to install daemon sets that were scheduled on it.

**Comment by thomastaylor312** [comment]
Did #188 possibly fix this issue for you?

**Comment by empath-75** [comment]
Yep, that was in my PR that I was waiting on approval to submit -- there's also the issue that it can't properly parse the error returned from the standard registry that he's describing because the details field is missing in the result returned from OciError needs to have the `details` field be an Option<>.  If you'll notice, he didn't see it in the krustlet logs, he saw it in the registry logs, because Krustlet can't parse the error response...

**Comment by empath-75** [comment]
You can close this, though -- thank you.
